### PR TITLE
Cache Storage to avoid allocate pinned memory for every prefill step and every decode step

### PR DESCRIPTION
**Pull request by dezhiAmd**
- Why
Before this change, storage in pinned memory returned by **for_transfer** are allocated every prefill step and every decode step. Cache storage can significantly reduce the allocation and release of pinned memory thus reduce the latency and improve the through-put

- How
Since logits are discarded every time after prefill step and decode step (and all other storage returned by for_transfer), cache them. The cache including prefill cache and decode cache is created when LlmGenerateService is created.

- Result (Single Server, SPX mode MI300X) 
 
Using script from https://github.com/nod-ai/benchmark_scripts

```
Batch size 8:
Before: "throughput": 0.9365385081863682, "latency_mean": 8.524904807806015
After:   "throughput": 0.953148448882024, "latency_mean": 8.375769250392914
 
Batch size 16:
Before:  "throughput": 1.0481085278749398, "latency_mean": 14.982843790054321
After:    "throughput": 1.1322175557670333, "latency_mean": 13.83236265540123
```

Using internal script:
```
Before:
Throughput Results:
Number of concurrent clients: 6
Total requests completed: 30
Test duration: 22.25 seconds
Average latency: 4.4514s
Min latency: 4.3440s
Max latency: 4.6043s
Throughput: 1.35 requests/second

After:
Throughput Results:
Number of concurrent clients: 6
Total requests completed: 30
Test duration: 20.91 seconds
Average latency: 4.1655s
Min latency: 4.0274s
Max latency: 4.2539s
Throughput: 1.43 requests/second
```

### CODE REVIEW

**Comment by stbaione** [comment]
What were the parameters of the benchmark that yielded 30%?

**Comment by dezhiAmd** [comment]
> What were the parameters of the benchmark that yielded 30%?

I followed the playbook

File: shortfin/python/shortfin_apps/llm/components/batcher.py [hunk 1]
```diff
@@ -591,3 +649,30 @@ async def get_results(self, logits, indices, req_count):
...
+        )
+
+        if not transfer:
+            return buffers
+
+        logits, indices = buffers
+
+        # Lazily initialize host buffers only once
+        if DecodeExecutorProcess._host_logits_init == 0:
+            DecodeExecutorProcess._host_logits = logits.for_transfer()
```
Comment (rsuderman) [inline shortfin/python/shortfin_apps/llm/components/batcher.py:665]: This is not safe. Transfer buffers between invocations are not guaranteed to be the same size.

File: shortfin/python/shortfin_apps/llm/components/batcher.py [hunk 2]
```diff
@@ -375,6 +357,14 @@ async def run(self):
 class PrefillExecutorProcess(LlmExecutorProcess):
     """Executes a prefill batch."""
 
+    tokens_host_cache: dict[tuple[int, int], sfnp.device_array] = {}
```
Comment (rsuderman) [inline shortfin/python/shortfin_apps/llm/components/batcher.py:360]: These would be globally shared between all executor processes with no cleanup. If we stopped the service then restarted it, then it would have invalid contents. We should not have static global state carried between processes as this would cause very bad outcomes in multi-instance.

**Review by rsuderman — CHANGES_REQUESTED** [review]

**Comment by stbaione** [comment]
> > What were the parameters of the benchmark that yielded 30%?
> 
> I followed the playbook

Was it with the internal script or the chai script? If chai script, which batch size?

File: shortfin/python/shortfin_apps/llm/components/batcher.py (hunk 1 above)
Comment (dezhiAmd) [reply shortfin/python/shortfin_apps/llm/components/batcher.py]: You are absolutely right. Logits size is determined by [bs, bsl], I have replaced a single logits_cache with a dictionary of logits which use the shape [bs, bsl] as the key to retrieve from the cache. The performance improvement can remain and now it is safe
